jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketChannelState.java

```diff
@@ -19,140 +19,223 @@
+    private State _channelState = State.CONNECTING;
+    private byte _incomingSequence = OpCode.UNDEFINED;
+    private byte _outgoingSequence = OpCode.UNDEFINED;
+    CloseStatus _closeStatus = null;
 
-    private AtomicReference<State> state = new AtomicReference<>(CONNECTING);
 
     public void onConnected()
     {
-        if (!state.compareAndSet(CONNECTING, CONNECTED))
-            throw new IllegalStateException(state.get().toString());
+        synchronized (this)
+        {
+            if (!_channelState.equals(State.CONNECTING))
+                throw new IllegalStateException(_channelState.toString());
+
+            _channelState = State.CONNECTED;
+        }
     }
 
     public void onOpen()
     {
-        if (!state.compareAndSet(CONNECTED, OPEN))
-            throw new IllegalStateException(state.get().toString());
+        synchronized (this)
+        {
+            if (!_channelState.equals(State.CONNECTED))
+                throw new IllegalStateException(_channelState.toString());
+
+            _channelState = State.OPEN;
+        }
     }
 
     @Override
     public String toString()
     {
-        return state.get().toString();
+        return _channelState.toString();
+    }
+
+
+    public State getState()
+    {
+        synchronized (this)
+        {
+            return _channelState;
+        }
     }
 
     public boolean isClosed()
     {
-        State s = state.get();
-        return !s.inOpen && !s.outOpen;
+        return getState().isClosed();
     }
 
     public boolean isInOpen()
     {
-        return state.get().inOpen;
+        return getState().isInOpen();
     }
 
     public boolean isOutOpen()
     {
-        return state.get().outOpen;
+        return getState().isOutOpen();
     }
 
     public CloseStatus getCloseStatus()
     {
-        return state.get().closeStatus;
+        synchronized (this)
+        {
+            return _closeStatus;
+        }
     }
 
-    public boolean onCloseIn(CloseStatus closeStatus)
+
+    public boolean onClosed(CloseStatus closeStatus)
     {
-        while (true)
+        synchronized (this)
         {
-            State s = state.get();
+            if (_channelState.equals(State.CLOSED))
+                return false;
 
-            if (!s.inOpen)
-                throw new IllegalStateException(state.get().toString());
+            _closeStatus = closeStatus;
+            _channelState = State.CLOSED;
+            return true;
+        }
+    }
 
-            if (s.outOpen)
-            {
-                State closedIn = new State("ICLOSED", false, true, closeStatus);
-                if (state.compareAndSet(s, closedIn))
-                    return false;
-            }
-            else
+
+    public boolean checkOutgoing(Frame frame) throws ProtocolException
+    {
+        synchronized (this)
+        {
+            _outgoingSequence = getNextState(frame, _outgoingSequence);
```

Comment:
Can you get the opcode once and then pass it into getNextState and use for the check below.   I think getOpCode is not just a trivial getter, but does a bitwise operation.